src/components/Toaster/ToasterSingleton.tsx

```diff
@@ -22,95 +14,86 @@ declare global {
+            },
+        ];
 
-        window[TOASTER_KEY] = this;
+        this.notify();
     }
 
-    destroy() {
-        // eslint-disable-next-line react/no-deprecated
-        ReactDOM.unmountComponentAtNode(this.rootNode);
-        document.body.removeChild(this.rootNode);
+    remove(name: string) {
+        this.toasts = removeToast(this.toasts, name);
+
+        this.notify();
     }
 
-    add = (options: ToastProps) => {
-        this.componentAPI?.add(options);
-    };
+    removeAll() {
+        this.toasts = [];
 
-    remove = (name: string) => {
-        this.componentAPI?.remove(name);
-    };
+        this.notify();
+    }
 
-    removeAll = () => {
-        this.componentAPI?.removeAll();
-    };
+    update(name: string, overrideOptions: Partial<ToastProps>) {
+        if (!hasToast(this.toasts, name)) {
+            return;
+        }
 
-    update = (name: string, overrideOptions: Partial<ToastProps>) => {
-        this.componentAPI?.update(name, overrideOptions);
-    };
+        const index = getToastIndex(this.toasts, name);
 
-    has = (name: string) => {
-        return this.componentAPI?.has(name) ?? false;
-    };
+        this.toasts = [
+            ...this.toasts.slice(0, index),
+            {
+                ...this.toasts[index],
+                ...overrideOptions,
+            },
+            ...this.toasts.slice(index + 1),
+        ];
 
-    private createRootNode() {
-        this.rootNode = document.createElement('div');
-        this.setRootNodeClassName();
-        document.body.appendChild(this.rootNode);
+        this.notify();
     }
 
-    private createReactRoot() {
-        if (ReactDOMClient) {
-            this.reactRoot = ReactDOMClient.createRoot(this.rootNode);
-        }
+    has(name: string) {
+        return hasToast(this.toasts, name);
     }
 
-    private render() {
-        const container = (
-            <ToasterProvider
-                ref={(api) => {
-                    this.componentAPI = api;
-                }}
-            >
-                <ToasterComponent hasPortal={false} mobile={this.mobile} />
-            </ToasterProvider>
-        );
-
-        if (this.reactRoot) {
-            this.reactRoot.render(container);
-        } else {
-            // eslint-disable-next-line react/no-deprecated
-            ReactDOM.render(container, this.rootNode, () => Promise.resolve());
+    subscribe(listener: (toasts: InternalToastProps[]) => void) {
+        if (typeof listener === 'function') {
+            this.listeners.push(listener);
         }
+
+        return () => {
+            this.listeners = this.listeners.filter(
+                (currentListener) => listener !== currentListener,
+            );
+        };
     }
 
-    private setRootNodeClassName() {
-        this.rootNode.className = bToaster({mobile: this.mobile}, this.className);
+    private notify() {
+        for (const listener of this.listeners) {
+            listener(this.toasts);
+        }
     }
```

Comment:
Please separate subscribe pattern from Toaster class to separate base class `EventEmitter`